### PR TITLE
[#64] Re-implemented `decimalP`

### PR DESCRIPTION
**Pull request by MitchStevens**
Re-implemented `decimalP`, disallowing leading zeros and double underscores to be matched when parsing integers. This is a copy of my previous pull request but from the master branch.

### CODE REVIEW

File: src/Toml/Parser.hs [hunk 1]
```diff
@@ -160,15 +161,30 @@ tableNameP :: Parser Key
 tableNameP = lexeme $ between (char '[') (char ']') keyP
 
 -- Values
-
 decimalP :: Parser Integer
-decimalP = mkInteger <$> decimalStringP
+decimalP = matchInteger >>= validateInteger
   where
-    decimalStringP   = fst <$> match (some digitChar >> many _digitsP)
```
Comment (ChShersh) [inline src/Toml/Parser.hs]: I want to say that I really like the previous implementation with `match` and `try`. I think it was quite efficient. Your implementation uses list of `Char`, but `match` relies on the efficient implementation of the underlying stream and uses of offsets. So I would like to have the implementation that behaves like the old one. But in this case it's kinda tricky to test, that the number is not started with zero but also not zero itself. For example, something like the following wouldn't work (but very close to the solution):

```haskell
mkInteger <$> notFollowedBy (some $ char '0') decimalStringP
```

**Comment by ChShersh** [comment]
@MitchStevens Do you need any help with this? Also, note that the `Toml.Parser` module was refactored recently and you need to add your changes now to another module.

**Comment by MitchStevens** [comment]
Sorry I haven't been looking at it lately, got a little distracted. I will have a look at how to write more efficient parsers and the new directory structure over the next couple of days.

**Comment by ChShersh** [comment]
@MitchStevens Don't worry! Feel free to ask any questions is something is not clear :slightly_smiling_face:

**Pull request closed without merging**